ihatemoney/templates/statistics.html

```diff
@@ -33,6 +33,243 @@ <h2>{{ _("Expenses by Month") }}</h2>
         {% endfor %}
         </tbody>
         </table>
+        <h2> {{ _("Visualization") }}</h2>
+        
+        
+        <!-- Required for some style in charts, feel free to move from here -->
+        <style>
+        #balanceBar .bb-ygrid-line line {stroke: red;}
+
+        #balanceBar .fill_green { fill: green; }
+        #balanceBar .fill_red { fill: red; }
+        </style>
+        <br>
+        <div id="balanceBar"></div>
+        <br>
+        <div id="memberPerMonth"></div>
+        <br>
+        <div id="combinedPerMonth"></div>
+        <br>
+        <div id="expPie"></div>
+        
+
+
+        
+        
+
+        
+
+        <script> 
+        var members = [];
+        var total_per_member = ['CurrentBalance'];
+        var total_exp_per_member = {};
+        var balance_min_value = Number.POSITIVE_INFINITY;
+        var balance_max_value = Number.NEGATIVE_INFINITY;
+        {% for stat in members_stats %}
```

Comment:
I'm not sure we actually want to loop here on the template. What do you think about passing the member_stats as json and using it on the javascript function, rather than looping on the template?

Reply:
I agree with this. Honestly, I am not too familiar with JavaScript so I am unsure of how I would go ahead and implement this. Any suggestions? I'll gladly implement it.

Reply:
Neat, thanks for the answer.

I can tell you the way I would do this, or at least how I would start : 

1. Remove the JavaScript code from the template ;
2. Create a separate file, import it on the template with a script tag ;
3. Try to limit the calls on the template to function call, which could pass JSON objects coming from the template
4. Split the JS logic into blocks of small bits to make it easier to grasp, do the iteration this way.

Hope it makes sense / helps. Let me know otherwise

Reply:
Ok, yeah this does make sense. I'm thinking about doing `onLoad` to call the function. But I am unsure on where I would call this. Since we want to limit the calls it probably is not a good idea to call it  on each chart `div` so would calling it on `<div class="d-flex flex-column">` be appropriate?

And more importantly, for passing the object into the function how would this be? Would it be like `myFunc({{ member_stats, months}})`?

Thanks!

Reply:
`onLoad` seems fine to me, as long as it calls when all the dependencies are loaded (wasn't clear from the docs).

I think one call for each graph would be just fine, so one fonction which would call and bind all the divs. You don't really need to group them all.

For passing the args, I believe you would need to convert them to JSON first. I would use a `|tojson` Jinja2 filter for this, so something like

```jinja2
javascriptfunction({{ member_stats | tojson }}, {{ months | tojson }})
```

Reply:
Sounds good. I'll implement this as soon as I have time, most likely during the weekend. Will update you if I run into any roadblocks, thanks for the help!

Reply:
Hello, so I'm implementing this but I've run into some early roadblocks. These deal with the fact that some of the stuff done in the Jinja template is dealing with python DateTime objects, which is pretty simple in Jinja. However, since I am putting this into another file I can't deal with it the same way. Specifically for `months` and `member_stats[monthly_exp]` since both deal with DateTime objects. For the latter, I can estimate that it might not be too difficult since I wrote the function that computes it, and it probably is sufficient to leave it as a string. 

However, this is not the case for `months`, and I do not want to change this as I am unsure of how many of the already implemented functions and HTML files expect `months` to behave in the way that it does. 

Any suggestions?

I've already made the other changes not related to moving the script to a separate file but I've not committed them.

Reply:
Hi, thanks for the heads up. If things are becoming too complex on the template side, that probably means that the data you send there is not in the right format, or is not prepared enough.

You might want to prepare it more, via some logic on the model or on the "views" /  "controllers". If that seems too complex, you can loop a bit on the templates for JS logic, but I think that's an anti-pattern generally speaking.

I hope that makes sense. Let me know if I can help you more.

Reply:
Yes, I understood that. However, that is not exactly what is causing the roadblock.

Right now, the only roadblock is that there is some data that is passed into the template as a Python DateTime object. The first of these is on the function that I wrote on models.py `get_monthly_expenditure` where the tuples in the list have types (DateTime, Double). This is not a huge issue because I could make it so instead of DateTime it is a string.

However, the more important one is the data that is passed to the template through the parameter `months` on `web.py`. The reason why this is more important is that it is used on every chart except the pie chart and I would have to change this to strings, and I think this is not a good idea because I am unsure of how many other templates or functions would be affected by changing the type in `months`.

What I can do, as you say, is possibly loop on the templates to convert `months` into an array of strings in JS and then pass that to the new JS function. Would that be an ok approach? I know it can be anti-pattern

Reply:
I'm not sure to fully understand what the problem is. I get the first part I think : that the data is passed as python datetime objects, but I'm not sure why it's an issue. 

From my perspective, you should let it like that and transform it to a string on the template side, if that's what you need. But maybe I'm missing something?

Reply:
Hello, thanks for the help! 

I was able to fix that issue and it all works perfectly now.

However, a possibly small issue has popped up. In order to JSONify the data passed into the JS function, I had to change `member_stats` on `models.py` such that instead of "member" being the person object, it was just the name because that is all that is used on statistics.html and moreover because if I didn't, the interpreter would throw an error that a Person object can't be serialized.

The issue that stems from this is that there are 2 tests that fail when I do this because the tests are set up to expect the whole Person object, not just the name. I know for a fact that this is the reason the tests fail because when I revert it, the only test that fails is one in `budget_test` that fails because the interpreter complained as I mentioned above.

Specifically, the failing test cases are: `test_statistics` on `budget_test` and `api_test`.

How would you like to proceed? I can easily change one of the failing tests (the one on `api_test`) so that now it accepts on just the name, but the other test that fails throws a strange HTML/Regex error that I am unsure how to change.

Reply:
Hey.

If it's not possible to serialize directly, you might want to add a `serialize` method which returns a python dict with serialized values, and call this in the template. If possible, I believe we shouldn't change the signature of the API unless that's really what we want. Hope it helps.

Reply:
Actually, there is already a way to serialize the model, see https://github.com/spiral-project/ihatemoney/blob/master/ihatemoney/models.py#L410

Reply:
If this doesn't make sense, you might want to push your code to this branch so i can have a look at it :-)

Reply:
Made sense. I made the changes and passed all tests. Pushed the changes to this branch